packages/flutter/lib/src/widgets/magnifier.dart

```diff
@@ -0,0 +1,545 @@
+      // from seeing its own styling.
+        Opacity(
+          opacity: decoration.opacity,
+          child: _MagnifierStyle(
+              decoration,
+              size: size,
+          ),
+        )
+      ],
+    );
+  }
+}
+
+class _MagnifierStyle extends StatelessWidget {
+  const _MagnifierStyle(this.decoration, {required this.size});
+
+  final MagnifierDecoration decoration;
+  final Size size;
+
+  @override
+  Widget build(BuildContext context) {
+    double largestShadow = 0;
+    for (final BoxShadow shadow in decoration.shadows ?? <BoxShadow>[]) {
+      largestShadow = math.max(
+          largestShadow,
+          (shadow.blurRadius + shadow.spreadRadius) +
+              math.max(shadow.offset.dy.abs(), shadow.offset.dx.abs()));
+    }
+
+    return ClipPath(
+      clipBehavior: Clip.hardEdge,
+      clipper: _DonutClip(
+        shape: decoration.shape,
+        spreadRadius: largestShadow,
+      ),
+      child: DecoratedBox(
+        decoration: decoration,
+        child: SizedBox.fromSize(
+          size: size,
+        ),
+      ),
+    );
+  }
+}
+
+/// A `clipPath` that looks like a donut if you were to fill its area.
+///
+/// This is necessary because the shadow must be added after the magnifier is drawn,
+/// so that the shadow does not end up in the magnifier. Without this clip, the magnifier would be
+/// entirely covered by the shadow.
+///
+/// The negative space of the donut is clipped out (the donut hole, outside the donut).
+/// The donut hole is cut out exactly like the shape of the magnifier.
+class _DonutClip extends CustomClipper<Path> {
+  _DonutClip({required this.shape, required this.spreadRadius});
+
+  final double spreadRadius;
+  final ShapeBorder shape;
+
+  @override
+  Path getClip(Size size) {
+    final Path path = Path();
+    final Rect rect = Offset.zero & size;
+
+    path.fillType = PathFillType.evenOdd;
+    path.addPath(shape.getOuterPath(rect.inflate(spreadRadius)), Offset.zero);
+    path.addPath(shape.getInnerPath(rect), Offset.zero);
+    return path;
+  }
+
+  @override
+  bool shouldReclip(_DonutClip oldClipper) => oldClipper.shape != shape;
+}
+
+class _Magnifier extends SingleChildRenderObjectWidget {
+  /// Construct a [_Magnifier].
+  const _Magnifier({
+      super.child,
+      required this.shape,
+      this.magnificationScale = 1,
+      this.focalPoint = Offset.zero,
+  });
+
+  /// [focalPoint] is the area the center of the
+  /// [_Magnifier] points to, relative to the center of the magnifier.
+  /// If left as [Offset.zero], the magnifier will magnify whatever is directly
+  /// below it.
+  final Offset focalPoint;
+
+  /// The scale of the magnification.
+  ///
+  /// A [magnificationScale] of 1 means that the content magi
+  final double magnificationScale;
+
+  /// The shape of the magnifier is dictated by [clip], which clips
+  /// the magnifier to the shape. If null, the shape will be rectangular.
```

Comment:
Good catch, this comment did not survive a refactor. Shape has clippers with shape.getOuterPath, I was referring to that, I'll elaborate here. I moved the nullability into MagnifierDecoration, so that comment does not belong anymore